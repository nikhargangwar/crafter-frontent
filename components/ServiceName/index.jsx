import React, { useEffect } from 'react';
import PropTypes from 'prop-types';
import { TextField } from '@mui/material';
import DoneIcon from '@mui/icons-material/Done';
// import { makeRequest } from '../../utils';
// import { apiEndPoints } from '../../constants';
// import { useParams } from 'react-router';

function ServiceName({ service, handleNameChange, serviceValue, setServiceValue, isEditable, setIsEditable}) {
  // const [isEditable, setIsEditable] = useState(false);
  // const [serviceValue, setServiceValue] = useState(service);
  // const token= localStorage.getItem('jwt_token');
  // const { blueprintId, version } = useParams();
  // const [isSaved, setIsSaved] = useState(false);

  useEffect(() => {
    setServiceValue(service);
  }, [service]);

  // const handleNameChange = async () => {
  //   setIsEditable(false);
  //   // setIsSaved(true);
  //   makeRequest.makeRequest(apiEndPoints.UPDATE_CONFIGURATIONS(blueprintId, serviceId, version), null , token, { data : { configurations: { name: { value: serviceValue}} }})
  //     .then(() => {
  //       setNodes((nds) =>
  //         nds.map((node) => {
  //           if (node.id === serviceId) {
  //             node.data = {
  //               ...node.data,
  //               label: serviceValue.toUpperCase(),
  //             };
  //           }
  //           return node;
  //         })
  //       );})
  //     .catch(() => {
  //       setServiceValue(service);
  //       setSnackbar({
  //         status: true,
  //         message: 'Service name already exists',
  //         type: 'error'
  //       });
  //     });
  // };

  return (
    <div className='service-element-border' onDoubleClick={() => setIsEditable(true)}>
      {isEditable ? (
        <TextField
          autoFocus
          className='w-[90%]'
          size='large'
          sx={{ fieldset: { borderColor: '#18CCCC'} ,input: { width:'100%', color: 'text.contrastText' ,border:'none',helperText:{color:'text.primary'}} }}
          value={ serviceValue.toLowerCase() }
          // onBlur = {() => {
          //   setIsEditable(false);
          //   // if (!isSaved) {
          //   //   setServiceValue(service);
          //   // }
          // }}
          InputProps={{
            endAdornment: (
              <DoneIcon
                data-testid='done-icon'
                sx={{ color: 'text.primary', fontSize: 20, cursor: 'pointer' }}
                onClick={() => handleNameChange()}
              />
            )}}
          onKeyPress={(e) => {
            if (e.key === 'Enter') {
              handleNameChange();
            }
          }}
          onChange={(e) => {
            setServiceValue(e.target.value);
            // setIsSaved(false);
          }}
        />
      ) : (
        <div className='service-element'>{serviceValue}</div>
      )}
    </div>
  );
}

export default ServiceName;

ServiceName.propTypes = {
  service: PropTypes.string.isRequired,
  handleNameChange: PropTypes.func.isRequired,
  serviceValue: PropTypes.string.isRequired,
  setServiceValue: PropTypes.func.isRequired,
  isEditable: PropTypes.bool.isRequired,
  setIsEditable: PropTypes.func.isRequired
};