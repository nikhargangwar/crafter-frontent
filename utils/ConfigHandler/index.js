const ConfigHandler = (event,updateConfig,userConfig,setUserFeConfig)=>{
  if(!updateConfig || !setUserFeConfig) return;
  setUserFeConfig({
    ...userConfig,
    [updateConfig]:event.target.value
  });
};

export default ConfigHandler;