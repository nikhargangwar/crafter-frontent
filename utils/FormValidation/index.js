import validateEmail from './emailValidation';
import validatePassword from './passwordValidation';
import validateString from './stringValidation';
import validateInteger from './integerValidation';

export default { validateEmail, validatePassword , validateString, validateInteger};
