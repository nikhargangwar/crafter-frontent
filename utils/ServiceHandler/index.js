export  let REPOSITORY ={
  username:'',
  email:'',
  repositoryImageAddress : '',
  token:''
};
export const customReactFlow = {
  left:0,
  top:84,
};

let FRONTEND = {
  port:'',
  numberOfReplicas : '',
  name:''
};
let BACKEND = {
  port:'',
  numberOfReplicas:'',
  name:''
};
let DATABASE={
  port:'',
  numberOfReplicas:'',
  name:'',
  dbUser:'',
  dbPassword:'',
  schemaName :''
};

export function getSpaceSeparated(configurations){
  let spaceSeperatedConfig=  Object.keys(configurations).reduce((acc,configuration)=>{
    let spaceSeperated = configuration
      .replace(/([A-Z])/g,' $1')
      .split(' ')
      .map(word => word && word[0].toUpperCase() + word.slice(1))
      .join(' ');
    acc.push(spaceSeperated);
    return acc;
  },[]);
  return spaceSeperatedConfig;
}
export function camelize(str) {
  return str.replace(/(?:^\w|[A-Z]|\b\w)/g, function(word, index) {
    return index === 0 ? word.toLowerCase() : word.toUpperCase();
  }).replace(/\s+/g, '');
}

export function getConfiguration(service){
  switch(service.split(' ')[0]){
  case 'FRONTEND':
    return FRONTEND;
  case 'BACKEND':
    return BACKEND;
  case 'DATABASE':
    return DATABASE;
  }
}

export function updateConfiguration(service,configuration){
  if(!service){
    REPOSITORY=configuration;
  }
  switch(service.split(' ')[0]){
  case 'FRONTEND':
    FRONTEND=configuration;
    break;
  case 'BACKEND':
    BACKEND=configuration;
    break;
  case 'DATABASE':
    DATABASE=configuration;
  }
}

export function getCustomEvent(serviceName){
  return {
    dataTransfer:{
      getData:(key)=>{
        if(key === 'application/reactflow'){
          return 'microservice';
        }else if(key === 'nameOfNode'){
          return serviceName;
        }
      }
    },
    clientX: getRandomCoordinates(-300,600,3),
    clientY: getRandomCoordinates(-300,600,3)
  };
}


export function getRandomCoordinates(from, to, fixed) {
  return (Math.random() * (to - from) + from).toFixed(fixed) * 1;
}

export function convertKibanaToCapital(str) {
  let words = str.split('_');
  let capitalizedWords = words.map(word => word.charAt(0).toUpperCase() + word.slice(1));
  return capitalizedWords.join(' ');
}

export function convertCamelCaseToCapital(str) {
  let words = str.split(/(?=[A-Z])/);
  let capitalizedWords = words.map(word => word.charAt(0).toUpperCase() + word.slice(1));
  return capitalizedWords.join(' ');
}

export function separatePascalCase(str) {
  return str?.split(/(?=[A-Z])/).join(' ');
}

export function capitalizeEachWord(str) {
  return str?.split(' ').map(word => word.charAt(0).toUpperCase() + word.slice(1)).join(' ');
}
