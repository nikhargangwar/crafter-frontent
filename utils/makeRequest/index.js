import axios from 'axios';
// import { apiEndPoints } from '../../constants/';
import { ERROR_ROUTE } from '../../constants/routes';

const serviceMap={
  'FRONTEND':'FrontEnd',
  'BACKEND':'BackEnd',
  'DATABASE':'Database',
};

function getRequestBody(imageTokens){
  const services=[];
  const mapServiceName = {};
  const validServices = ['FRONTEND','BACKEND','DATABASE'];
  for (const key in localStorage) {
    const service_type = key.split(' ')[0];
    if(validServices?.indexOf(service_type) !== -1){
      const configuration = localStorage.getItem(key);
      let service={};
      service['service_type'] = serviceMap[service_type];
      service['configurations'] = JSON.parse(configuration).userConfig;
      service['customEnv']= JSON.parse(configuration).customConfig;
      service['connected_service']=[];
      console.log(service['customEnv']['offline_enabled']);
      if(!process.env.OFFLINE_ENABLED)
      {
        service['imageRepository'] = JSON.parse(configuration).dockerConfig;
        service['imageRepository']['token'] = imageTokens[key];
      }
      const serviceName = JSON.parse(configuration).userConfig.name;
      mapServiceName[key] = [serviceName,services.length];
      services.push(service);
    }
  }
  const connections =  JSON.parse(localStorage.getItem('connection'));
  const newConnections = {};
  if(connections){
    let index = 0;
    Object.keys(connections).map(service => {
      index = mapServiceName[service][1];
      newConnections[mapServiceName[service][0]]=[];
      connections[service].map(connected_service=>{
        newConnections[mapServiceName[service][0]].push(mapServiceName[connected_service][0]);
      });
      services[index]['connected_service'] = newConnections[mapServiceName[service][0]];
    });
  }
  return services;
}

const makeRequest = async (
  apiEndPoint,
  navigate,
  token,
  dynamicConfig = {}
) => {
  try {
    const requestDetails = {
      baseURL: apiEndPoint.baseURL,
      url: apiEndPoint.url,
      method: apiEndPoint.method,
      headers: {
        Authorization: token
      },
      ...dynamicConfig,
    };
    const {data} = await axios(requestDetails);
    return data;
  } catch (error) {
    if(navigate)
    {
      const errorStatus = error.response?.status;
      if (!errorStatus) {
        navigate(ERROR_ROUTE);
      } else {
        navigate(`${ERROR_ROUTE}/${errorStatus}`);
      }
      return null;
    }
    else{
      throw new Error(error);
    }
    
  }
};

const authMakeRequest = async (configObj) => {
  const { data } = await axios(configObj);
  return data;
};

export default { makeRequest, authMakeRequest, getRequestBody };
