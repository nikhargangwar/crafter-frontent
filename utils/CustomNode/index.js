/* eslint-disable */
import React, { memo } from 'react';
import './CustomNode.css';
import { Handle, Position } from 'reactflow';
import propTypes from 'prop-types';
import reactIcon from '../../assets/REACT-3d.png';
import nodeIcon from '../../assets/EXPRESS-3d.png';
import postgresIcon from '../../assets/postgres-3d.png';
import { useEffect } from 'react';
import { useState } from 'react';
import makeRequest from '../makeRequest';
import { apiEndPoints } from '../../constants';
import { useNavigate } from 'react-router-dom';

// eslint-disable-next-line react/display-name, react/prop-types
export default memo(({ data}) => {
  const service = data.label.split(' ')[0];
  const pluginId = data.pluginId;
  const navigate = useNavigate();

  const [imgSrc,setImgSrc] = useState();
useEffect(()=>{
  makeRequest.makeRequest(apiEndPoints.GET_PLUGIN_BY_ID(pluginId), navigate, localStorage.getItem('jwt_token'))
  .then((response)=>{
    // setBlueprintName( response.metadata.name);
    console.log('hjhj',response.metadata.icon_node_filename);
    setImgSrc(response.metadata.icon_node_filename);
  });

},[imgSrc])
  // useEffect(() => {
  //   switch (service) {
  //     case 'REACT-FRONTEND':
  //       setImgSrc(reactIcon);
  //       break;
  //     case 'EXPRESS-BACKEND':
  //       setImgSrc(nodeIcon);
  //       break;
  //     case 'POSTGRESQL-DATABASE':
  //       setImgSrc(postgresIcon);
  //       break;
  //     default:
  //       setImgSrc(postgresIcon);
  //       break;
  //   }
  // }, [imgSrc]);

  return (

  <div className='node'>
      <div className="node-handles">
            <Handle type="source" style = {{width:'10px',height:'10px' }} position={Position.Top} id="a" />
            <Handle type="source" style = {{width:'10px',height:'10px' }}position={Position.Right} id="b" />
            <div className="node-content">
    
            <Handle type="source" style = {{width:'10px',height:'10px' }}position={Position.Bottom} id="c" />
            <Handle type="source" style = {{width:'10px',height:'10px' }}position={Position.Left} id="d" />
            <img src={`${process.env.REACT_APP_API_GATEWAY_URL}/minio/images/${imgSrc}`} alt="microservice-icon" />
       
       </div>
      </div>
      <div className="node-name">
      {
            data.label
          }
      </div>
      
  </div>
   

  
  );
});

