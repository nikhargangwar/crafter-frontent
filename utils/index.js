import CustomNode from './CustomNode';
import FormValidation from './FormValidation';
import makeRequest from './makeRequest';
import {camelize, getConfiguration, getSpaceSeparated, updateConfiguration} from './ServiceHandler';
export { CustomNode, FormValidation, makeRequest, camelize, getConfiguration, getSpaceSeparated, updateConfiguration };