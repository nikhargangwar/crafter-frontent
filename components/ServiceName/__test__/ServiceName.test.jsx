import { fireEvent, render, screen } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import { BrowserRouter as Router } from 'react-router-dom';
import ServiceName from '..';
import { makeRequest } from '../../../utils';

jest.mock('../../../utils', () => ({
  ...jest.requireActual('../../../utils'),
  makeRequest: {
    makeRequest: jest.fn(),
  },
}));

describe('ServiceName component', () => {
  it('should render correctly', () => {
    const { asFragment } = render(<Router>
      <ServiceName 
        service="Test Service"
        serviceId="123"
        setNodes={() => {}}
        serviceValue="Test Service"
        setServiceValue={jest.fn()}
        setIsEditable={jest.fn()}
      />
    </Router>);
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render the service name', async () => {
    render(<Router>
      <ServiceName 
        service="Test Service"
        serviceId="123"
        setNodes={() => {}}
        serviceValue="Test Service"
        setServiceValue={jest.fn()}
        setIsEditable={jest.fn()}
      />
    </Router>);
    const serviceName = await screen.findByText('Test Service');
    expect(serviceName).toBeInTheDocument();
  });

  it('should render the service name in textfield on double click', async () => {
    render(<Router>
      <ServiceName 
        service="Test Service"
        serviceId="123"
        setNodes={() => {}}
        serviceValue="Test Service"
        setServiceValue={jest.fn()}
        setIsEditable={jest.fn()}
        isEditable={false}
      />
    </Router>);
    const serviceName = await screen.findByText('Test Service');
    fireEvent.doubleClick(serviceName);
    render(<Router>
      <ServiceName 
        service="Test Service"
        serviceId="123"
        setNodes={() => {}}
        serviceValue="Test Service"
        setServiceValue={jest.fn()}
        setIsEditable={jest.fn()}
        isEditable={true}
      />
    </Router>);
    const textField = await screen.findByRole('textbox');
    expect(textField).toBeInTheDocument();
  });

  it('should change the service name on enter', async () => {
    const mockedHandleNameChange = jest.fn();
    render(<Router>
      <ServiceName 
        service="Test Service"
        serviceId="123"
        setNodes={() => {}}
        serviceValue="Test Service"
        setServiceValue={jest.fn()}
        setIsEditable={jest.fn()}
        isEditable={false}
        handleNameChange={mockedHandleNameChange}
      />
    </Router>);
    jest.spyOn(makeRequest, 'makeRequest').mockResolvedValue({});
    const serviceName = await screen.findByText('Test Service');
    fireEvent.doubleClick(serviceName);
    render(<Router>
      <ServiceName 
        service="Test Service"
        serviceId="123"
        setNodes={() => {}}
        serviceValue="Test Service"
        setServiceValue={jest.fn()}
        setIsEditable={jest.fn()}
        isEditable={true}
        handleNameChange={mockedHandleNameChange}
      />
    </Router>);
    const textField = await screen.findByRole('textbox');
    fireEvent.change(textField, { target: { value: 'New Service' } });
    fireEvent.keyPress(textField, { key: 'Enter', code: 'Enter', charCode: 13 });
    expect(mockedHandleNameChange).toHaveBeenCalledTimes(1);
  });

  it('should change the service name on click of done icon', async () => {
    const mockedHandleNameChange = jest.fn();
    render(<Router>
      <ServiceName 
        service="Test Service"
        serviceId="123"
        setNodes={() => {}}
        serviceValue="Test Service"
        setServiceValue={jest.fn()}
        setIsEditable={jest.fn()}
        isEditable={false}
      />
    </Router>);
    jest.spyOn(makeRequest, 'makeRequest').mockResolvedValue({});
    const serviceName = await screen.findByText('Test Service');
    fireEvent.doubleClick(serviceName);
    render(<Router>
      <ServiceName 
        service="Test Service"
        serviceId="123"
        setNodes={() => {}}
        serviceValue="Test Service"
        setServiceValue={jest.fn()}
        setIsEditable={jest.fn()}
        isEditable={true}
        handleNameChange={mockedHandleNameChange}
      />
    </Router>);
    const textField = await screen.findByRole('textbox');
    fireEvent.change(textField, { target: { value: 'New Service' } });
    const doneIcon = await screen.findByTestId('done-icon');
    fireEvent.click(doneIcon);
    expect(mockedHandleNameChange).toHaveBeenCalledTimes(1);
  });
});
