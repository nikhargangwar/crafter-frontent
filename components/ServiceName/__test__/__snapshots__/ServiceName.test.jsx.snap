// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ServiceName component should render correctly 1`] = `
<DocumentFragment>
  <div
    class="service-element-border"
  >
    <div
      class="service-element"
    >
      Test Service
    </div>
  </div>
</DocumentFragment>
`;
